contracts/src/StakeTable.sol

```diff
@@ -0,0 +1,149 @@
+    uint256[2] private totalStakeArr;
+    uint32 private totalNumberOfKeys;
+    uint256 private totalVotingStakeVal;
+    uint64 private numRegistrations;
+    uint64 private numPendingExits;
+    uint64 private constant BLOCKS_PER_EPOCH = 10; // TODO make an argument of the constructor?
+    uint256 private creationBlock;
+    address private tokenAddress;
+
+    /// @dev Computes a hash value of some G2 point
+    /// @param blsVK BLS verification key in G2
+    /// @return keccak256(blsVK)
+    function hashBlsKey(BN254.G2Point calldata blsVK) private pure returns (bytes32) {
+        uint256 x0 = blsVK.x0;
+        uint256 x1 = blsVK.x1;
+        uint256 y0 = blsVK.y0;
+        uint256 y1 = blsVK.y1;
+        bytes32 hash = keccak256(abi.encode(x0, x1, y0, y1));
+        return hash;
+    }
+
+    constructor(address _tokenAddress) {
+        creationBlock = block.number;
+        tokenAddress = _tokenAddress;
+    }
+
+    function currentEpoch() private view returns (uint64) {
+        return uint64((block.number - creationBlock) / BLOCKS_PER_EPOCH);
+    }
+
+    function totalStake() external view returns (uint256, uint256) {
+        return (totalStakeArr[0], totalStakeArr[1]);
+    }
+
+    function totalKeys() external view returns (uint32) {
+        return totalNumberOfKeys;
+    }
+
+    function totalVotingStake() external view returns (uint256) {
+        return totalVotingStakeVal;
+    }
+
+    function lookupStake(BN254.G2Point calldata blsVK) external view returns (uint64) {
+        Node memory node = this.lookupNode(blsVK);
+        return node.balance;
+    }
+
+    function lookupNode(BN254.G2Point calldata blsVK) external view returns (Node memory) {
+        bytes32 hash = hashBlsKey(blsVK);
+        return nodesTable[hash];
+    }
+
+    function nextRegistrationEpoch() external view returns (uint64) {
+        // TODO implement queue logic
+        return currentEpoch() + 1;
+    }
+
+    function numPendingRegistrations() external view returns (uint64) {
+        return numRegistrations;
+    }
+
+    function nextExitEpoch() external view returns (uint64) {
+        if (numPendingExits == 0) {
+            return 0;
+        } else {
+            return 1;
+        }
+    }
+
+    function numPendingExit() external view returns (uint64) {
+        return numPendingExits;
+    }
+
+    function register(
+        BN254.G2Point calldata blsVK,
+        EdOnBN254.EdOnBN254Point calldata schnorrVK,
+        uint64 amount,
+        StakeType stakeType,
+        BN254.G1Point calldata blsSig,
+        uint64 validUntilEpoch
+    ) external returns (bool) {
+        bytes32 key = hashBlsKey(blsVK);
+        Node memory node = nodesTable[key];
+
+        // The node must not already be registered.
+        require(node.account == address(0x0), "The node has already been registered");
+
+        bytes memory message = abi.encode(msg.sender);
+        BLSSig.verifyBlsSig(message, blsSig, blsVK);
```

Comment:
this is not a problem. There's no benefit for registering someone else's signature. Squatting that pk doesn't give the attacker any advantage. afterall, from the HotShot system perspective, you, the attacker is the new staker now, good. you are expected to vote (oops.. you don't have the secret key, slashed)

plus when you register, you are paying yourself (w/ your own token)

plus, this so-called "rushing attack" (or I prefer "PK squatter attack") is orthogonal to what message we sign. As long as there's mempool, people could do that (even tho there's no point doing so)

Reply:
Indeed the attack described above does not seem to do any harm, however allowing to sign the same message to  register a [rogue key](https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html ) as follows (For the ease of exposition we assume $\mathbb{G_1}=\mathbb{G_2 }=\mathbb{G}$ and $g$ generates $\mathbb{G}$):
1. An honest participant registers its public key $pk=g^{sk}$ where $sk$ is the signing key . To do so, he signs the unique message $m$ producing $\sigma=H(m)^{sk}$.
1. A malicious participant will now use $pk$ and $\sigma$ in order to register a rogue public key $pk^{\star}:=g^\beta pk^{-1}$, where $\beta$ is chosen by the attacker. To achieve this goal the attacker will need to compute a signature on the message $m$ that verifies against the rogue public key. He sets $\sigma^* := H(m)^\beta \sigma^{-1} = H(m)^\beta H(m)^{-sk}=H(m)^{\beta-sk}$. This signature will pass the BLS check as: $$e(H(m),pk^*)=e(H(m),g^\beta pk^{-1})=e(H(m),g^{\beta-sk})$$

Reply:
This is true, but again, how does this amounts to an attack? 
i agree that "BLS signature verification != proof of knowledge on the secret key", but I don't see how registering a rouge pk can do any good to the attacker.

Reply:
Once the rogue key $pk^{\star}$ is registered, the attacker can sign messages on the aggregate public key $pk^*pk=g^\beta pk^{-1}pk=g^\beta$.  In our context an attacker could register a rogue public key $g^\beta \Pi_{k=1}^N pk_k^{-1}$ so that the $N+1$ keys correspond to more than 2/3 of the stake and thus sign QCs on his own.

Reply:
you are right. we do need to sign distinct messages then.